[4.0] Add a parameter for "back-to-top" button in Cassiopeia

drmenzelit opened the pull request:
Pull Request for Issue #25239 .

### Summary of Changes
Added a parameter to show / hide the back-to-top button in Cassiopeia

### Testing Instructions

Install PR, run npm ci or compile the scss files from Cassiopeia
Go to template, under "advanced" a new switcher should appear
![Screenshot_2020-08-21 Templates Edit Style - Joomla Cassiopeia - Administration](https://user-images.githubusercontent.com/9153168/90889405-6a972780-e3b8-11ea-8906-962470595333.png)


### Actual result BEFORE applying this Pull Request

Currently the footer appears when a module has been published on position "footer". And the back-to-top only appears if the footer is visible. Since the footer is part of the grid, it is not easy to move the back-to-top outside the grid.


### Expected result AFTER applying this Pull Request

Now it is possible to decide to have a back-to-top or not. If yes, the footer will be rendered showing the button, independently if there is a module in the position footer.
It is also possible to show only modules (switch back-to-top off) or modules and back-to-top (switch on).

Example with module on footer and back-to-top button

![Screenshot_2020-08-21 Home](https://user-images.githubusercontent.com/9153168/90889516-9e724d00-e3b8-11ea-8a70-acd90a968b98.png)


### Documentation Changes Required

drmenzelit committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

drmenzelit committed:
Merge remote-tracking branch 'joomla-origin/4.0-dev' into 4.0-dev

drmenzelit committed:
[4.0] Add a parameter for back-to-top button in Cassiopeia

Added a parameter to show / hide the back-to-top button in Cassiopeia.
Currently the footer appears when a module has been published on position "footer". And the back-to-top only appears if the footer is visible. Since the footer is part of the grid, it is not easy to move the back-to-top outside the grid.

Now it is possible to decide to have a back-to-top or not. If yes, the footer will be rendered showing the button, independently if there is a module in the position footer.
It is also possible to show only modules (switch back-to-top off) or modules and back-to-top (switch on).

brianteeman commented on language/en-GB/tpl_cassiopeia.ini:
```diff
@@ -9,5 +9,6 @@ TPL_CASSIOPEIA_FLUID_LABEL="Fluid Layout"
 TPL_CASSIOPEIA_LOGO_LABEL="Logo"
 TPL_CASSIOPEIA_STATIC="Static"
 TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
+TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top button"
```
```suggestion
TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top Button"
```

To follow the style guidelines

drmenzelit replied:
Changed on new commit

drmenzelit replied:
Changed with new commit

drmenzelit committed:
Added smooth scrolling

Add smooth scrolling taking in account the preferences of the user (reduced motion).
Improve  language file.

drmenzelit committed:
Correct height and position of back-to-top button

drmenzelit committed:
Correct code style

drmenzelit committed:
Correct code style 2

drmenzelit committed:
Change smooth scrolling behavior to take in account reduced motion preferences

Quy commented on language/en-GB/tpl_cassiopeia.ini:
```diff
@@ -9,5 +9,6 @@ TPL_CASSIOPEIA_FLUID_LABEL="Fluid Layout"
 TPL_CASSIOPEIA_LOGO_LABEL="Logo"
 TPL_CASSIOPEIA_STATIC="Static"
 TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
+TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top Button"
```
Please sort key in alpha order.

drmenzelit replied:
Corrected on last commit

drmenzelit committed:
Language file corrected, language string changed to match others

brianteeman commented on templates/cassiopeia/scss/blocks/_global.scss:
```diff
@@ -1,4 +1,12 @@
+:root {
+  scroll-behavior: smooth;
+
+  // No smooth scroll for people who prefer not to
```
Suggesion

```suggestion
  // Disable smooth scroll when the user has set "prefers-reduced-motion"
```

brianteeman commented on language/en-GB/tpl_cassiopeia.ini:
```diff
@@ -4,6 +4,7 @@
 ; Note : All ini files need to be saved as UTF-8
 
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
+TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Button"
```
technically its a link not a button


```suggestion
TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
```

infograf768 commented on templates/cassiopeia/index.php:
```diff
@@ -173,16 +173,15 @@
 	<footer class="grid-child container-footer footer">
-		<hr>
-		<p class="float-right">
+		<jdoc:include type="modules" name="footer" style="none" />
```
Please test after installing blog sample data.
The menu module is next to the button, so much that they look related. This is really confusing.

drmenzelit committed:
Update language/en-GB/tpl_cassiopeia.ini

Link instead button

Co-authored-by: Brian Teeman <brian@teeman.net>

drmenzelit committed:
Update templates/cassiopeia/scss/blocks/_global.scss

Comment for smooth scroll

Co-authored-by: Brian Teeman <brian@teeman.net>

hans2103 commented:
I have tested this item :white_check_mark: successfully on baac7fb9d4bdf13af9f2b14a718d1b19bbda027b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

hans2103 commented:
print styling will be taken care of in redesign of Cassiopeia. George mentioned https://twitter.com/dgrammatiko/status/1298231873425092609?s=21 which is interesting <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

alikon commented:
i really would like to close this pointing to a real pr , can you  share a link ?

hans2103 commented:
> i really would like to close this pointing to a real pr , can you share a link ?

I hope this is good enough
https://github.com/joomla/cassiopeia/issues

chmst commented:
I have tested this item :white_check_mark: successfully on baac7fb9d4bdf13af9f2b14a718d1b19bbda027b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

infograf768 commented:
Tks.

### Reminder
The module displayed in the footer position is real bad and should be corrected.
<img width="1155" alt="Screen Shot 2020-08-26 at 10 49 04" src="https://user-images.githubusercontent.com/869724/91282885-33df5980-e78a-11ea-9e6b-d935be5bf54b.png">

Scrabble96 commented:
As I am unable to see the current Cassiopeia go top arrow at all (as of yesterday's Nightly update) - with or without a footer module published - I would like to tentatively suggest this solution for a go to top function that sits outside the footer. Its colours, size and position are controlled by css:

1. In Cassiopeia index.php add this code to below the footer module, above or below the debug:

`<div onClick="topFunction()" id="gotop" class="hide" title="<?php echo Text::_('TPL_CASSIOPEIA_BACKTOTOP'); ?>"><i class="fas fa-arrow-up arrow-up"></i></div>`

2. Add this js to template.js (or user.js for testing purposes):

`/*  Go to Top Link `
`-------------------*/`
`var topbutton = document.getElementById("gotop");`
`window.onscroll = function() {scrollFunction()};`
`function scrollFunction() {`
`  if (document.body.scrollTop > 450 || document.documentElement.scrollTop > 450) {`
`    document.getElementById("gotop").className = "top-link";`
`  } else {`
`    document.getElementById("gotop").className = "top-link hide";`
`  }`
`}`
`function topFunction() {`
`  document.body.scrollTop = 0; // For Safari`
`  document.documentElement.scrollTop = 0; // For Chrome, Firefox, IE and Opera`
`}`

3. Add this css:

`.top-link {`
`  visibility: visible;}`
`.hide {`
` display: none;}`
`.top-link .arrow-up {`
`  align-items: center;`
`  animation-duration: 1s;`
`  background: red;`
`  border: 1px solid white;`
`  bottom: 0;`
`  color: white;`
`  cursor: pointer;`
`  font-size: 20px;`
`  height: 45px;`
`  margin: 0 1em 0 0;`
`  padding-top: .5em;`
`  position: fixed;`
`  left: 0;`
`  text-align: center;`
`  width: 50px;`
`}`
`.top-link .arrow-up:hover {`
`  background: pink;`
`  padding-top: .35em;`
`}`

richard67 commented:
@Scrabble96 You know that you have to switch it on in Cassiopeia's template style settings in the "Advanced" tab?

Scrabble96 commented:
> 
> 
> @Scrabble96 You know that you have to switch it on in Cassiopeia's template style settings in the "Advanced" tab?

I do now! It wasn't there in Protostar, so it's going to have to be made obvious to new users of Cassiopeia that it is switched on or off in the Styles Advanced tab.

richard67 commented:
@Scrabble96 The description of this PR here explains that clearly. I assumed when you are commenting this PR then yopu have read its description, too. And sure, later it has to be described somewhere in documentations.

Scrabble96 commented:
> 
> 
> @Scrabble96 The description of this PR here explains that clearly. I assumed when you are commenting this PR then yopu have read its description, too. And sure, later it has to be described somewhere in documentations.

Sorry. Long day. I'll delete my comment.

richard67 commented:
Sorry if I sounded too harsh. Was a long day for me, too.
